### ISSUE TITLE
initialOpen={true} is not working.

### ISSUE THREAD

**Issue by nilesh2b** [minop1205/react-dnd-treeview#175]
@minop1205 Hi, I am using this package and it solves almost my all problems.
But I am not able to expand the dynamic node tree by using initialOpen attribute.
Thanks

**Comment by minop1205** [comment]
@nilesh2b Thanks for the report.
It would be helpful to know the code and steps to reproduce it.

Or can't you use the `open` / `close` methods instead of the `initialOpen` API?

**Reference below**

https://github.com/minop1205/react-dnd-treeview#usage-to-open--close-methods

https://minop1205.github.io/react-dnd-treeview/?path=/docs/basic-examples-opening-and-closing-all-nodes--opening-and-closing-all-nodes-story

**Comment by nilesh2b** [comment]
@minop1205 Thank you for your prompt response.
I wanted to open the tree on initial page load

**Comment by nilesh2b** [comment]
<DndProvider backend={MultiBackend} options={getBackendOptions()}>
                <div>{open && <AddDialog tree={treeData} onClose={handleCloseDialog} onSubmit={handleSubmit} />}</div>
                <Tree tree={treeData} rootId={"0"} onDrop={handleDrop} initialOpen={true} classes={{ container: "wtree", listItem: "parent-listitem" }} render={(node, { depth, isOpen, hasChild, onToggle }) => <CustomNode handleOpenDialog={handleOpenDialog} node={node} depth={depth} hasChild={hasChild} isOpen={isOpen} onToggle={onToggle} onTextChange={handleTextChange} onDelete={handleDelete} onCopy={handleCopy} dragPreviewRender={(monitorProps) => <CustomDragPreview monitorProps={monitorProps} />} />} />
              </DndProvider>
Please check the above code i have implemented.

**Comment by nilesh2b** [comment]
```
<DndProvider backend={MultiBackend} options={getBackendOptions()}>
  <div>
    {open && (
      <AddDialog
        tree={treeData}
        onClose={handleCloseDialog}
        onSubmit={handleSubmit}
      />
    )}
  </div>
  <Tree
    tree={treeData}
    rootId={"0"}
    onDrop={handleDrop}
    initialOpen={true}
    classes={{ container: "wtree", listItem: "parent-listitem" }}
    render={(node, { depth, isOpen, hasChild, onToggle }) => (
      <CustomNode
        handleOpenDialog={handleOpenDialog}
        node={node}
        depth={depth}
        hasChild={hasChild}
        isOpen={isOpen}
        onToggle={onToggle}
        onTextChange={handleTextChange}
        onDelete={handleDelete}
        onCopy={handleCopy}
        dragPreviewRender={(monitorProps) => (
          <CustomDragPreview monitorProps={monitorProps} />
        )}
      />
    )}
  />
</DndProvider>;
```

**Comment by nilesh2b** [comment]
The above mentioned code i have implemented.

**Comment by minop1205** [comment]
@nilesh2b 
Looking at the code, there doesn't seem to be anything strange about it.
It is odd that `dragPreviewRender` is a CustomNode props, but it does not seem to be related to the `initialOpen` issue.

The `initialOpen` API only handles the tree component when it is mounted, so it will not work on nodes added after it is mounted. If for any other reason it does not work as intended, please provide a repository or CodeSandbox that can reproduce the problem.

**Comment by nilesh2b** [comment]
Yes you are right, nodes are added after it is mounted. In that case how I could open the tree on initial load.

**Comment by nilesh2b** [comment]
`import React, { useState, useEffect, Suspense, useRef, useCallback } from "react"
import { useParams } from "react-router-dom"
import { useLocation } from "react-router-dom"
import "./ProcessCss.css"
import { Grid, Typography } from "@mui/material"
import Navbar from "../../components/Navbar/Navbar"
import Button from "@mui/material/Button"
import { Tree, getBackendOptions, MultiBackend, getDescendants } from "@minoru/react-dnd-treeview"
import { DndProvider } from "react-dnd"
import EditOutlinedIcon from "@mui/icons-material/EditOutlined"
import { ThemeProvider } from "@mui/material"
import { theme } from "../../theme"
import { useTheme } from "@mui/material/styles"
import { CustomNode } from "../../components/ProcessHierarchyCustomNode/CustomNode"
import { CustomDragPreview } from "../../components/ProcessHierarchyCustomNode/CustomDragPreview"
import { AddDialog } from "../../components/ProcessHierarchyCustomNode/AddDialog"
import { ProcessSave } from "../../components/ProcessHierarchyCustomNode/ProcessSave"

import Dialog from "@mui/material/Dialog"
import DialogActions from "@mui/material/DialogActions"
import DialogContent from "@mui/material/DialogContent"
import DialogContentText from "@mui/material/DialogContentText"
import DialogTitle from "@mui/material/DialogTitle"
import MenuItem from "@mui/material/MenuItem"
import OutlinedInput from "@mui/material/OutlinedInput"
import FormControl from "@mui/material/FormControl"
import Select from "@mui/material/Select"

import * as Constant from "../../comman/constant"
import { BASE_URL } from "../../comman/constant"

const ITEM_HEIGHT = 48
const ITEM_PADDING_TOP = 8
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 250,
    },
  },
}

const getLastId = (treeData) => {
  const reversedArray = [...treeData].sort((a, b) => {
    if (a.id < b.id) {
      return 1
    } else if (a.id > b.id) {
      return -1
    }
    return 0
  })
  if (reversedArray.length > 0) {
    return reversedArray[0].id
  }
  return 0
}
const ClientProcess = (props) => {
  const { id } = useParams()
  const [loading, setLoading] = useState(false)
  const [showBtn, setShowBtn] = useState(false)
  const [nodeData, setNodeData] = useState([])
  const [arrWith, setArrWith] = useState([])
  const [nodeDepth, setNodeDepth] = useState()

  const [openDialog, setOpenDialog] = useState(false)
  const theme = useTheme()
  const [personName, setPersonName] = useState([])
  const [textValue, setTextValue] = useState("")
  const location = useLocation()

  const handleClickOpen = () => {
    setOpenDialog(true)
  }

  const handleClose = () => {
    setOpenDialog(false)
  }
  const handleChange = (event) => {
    const {
      target: { value },
    } = event
    setPersonName(
      // On autofill we get a stringified value.
      typeof value === "string" ? value.split(",") : value
    )
  }

  const names = ["PDF", "Excel"]
  function getStyles(name, personName, theme) {
    return {
      fontWeight: personName.indexOf(name) === -1 ? theme.typography.fontWeightRegular : theme.typography.fontWeightMedium,
    }
  }

  // function onTextValueChange(e){
  //   setTextValue(e.target.value)
  // }

  const onTextValueChange = (event) => {
    const {
      target: { value },
    } = event
    setTextValue(
      // On autofill we get a stringified value.
      typeof value === "string" ? value.split(",") : value
    )
  }

  useEffect(() => {
    const loadPhLevels = async () => {
      await fetch(BASE_URL + Constant.PROCESS_HIERARCHY_TREE + `${id}`)
        .then((response) => response.json())
        .then((data) => {
          setNodeData(data)
        })
        .catch((err) => {
          console.log(err.message)
        })
      setLoading(false)
    }
    loadPhLevels()
  }, [])

  const newArr = useCallback((arrData) => {
    let arr0 = [delete arrData.industryId]
    const arr1 = [arrData].flatMap((elem) => elem.clientl2bphList)

    const arr2 = arr1
      .map((elem) => {
        if (elem) {
          return elem.clientl3bphList
        }
      })
      .filter((el) => {
        if (el) {
          return el != 0
        }
      })
      .map((e) => {
        if (e) {
          return e
        }
      })
      .flat(1)

    const arr2_4bph = arr2
      .map((elem) => {
        if (elem) {
          return elem.clientl4bphList
        }
      })
      .flat(2)

    const arr2_5bph = arr2_4bph
      .map((elem) => {
        if (elem) {
          return elem.clientl5bphList
        }
      })
      .flat(1)

    const simpleObj = {}
    const destructure = (obj) => {
      for (let key in obj) {
        const value = obj[key]
        const type = typeof value
        if (["string", "boolean"].includes(type) || (type === "number" && !isNaN(value))) {
          simpleObj[key] = value
        }
      }
      Object.assign(simpleObj, { parent: "0" })
      return simpleObj
    }

    arr1.push(destructure(arrData))
    arr2.push(arr1)
    arr2.push(arr2_4bph)
    arr2.push(arr2_5bph)

    const multiDimArray = arr2
      .map((items) => {
        if (items) {
          return items
        }
      })
      .flat(1)

    const renameDescription = JSON.parse(JSON.stringify(multiDimArray).split('"description":').join('"text":'))
    const renameParent1 = JSON.parse(JSON.stringify(renameDescription).split('"clientL1BPHId":').join('"parent":'))
    const renameParent2 = JSON.parse(JSON.stringify(renameParent1).split('"clientL2BPHId":').join('"parent":'))
    const renameParent2_1 = JSON.parse(JSON.stringify(renameParent2).split('"clientL3BPHId":').join('"parent":'))
    const renameParent2_2 = JSON.parse(JSON.stringify(renameParent2_1).split('"clientL4BPHId":').join('"parent":'))

    const arrWithDroppable = renameParent2_2.map((object) => {
      return { ...object, droppable: true }
    })
    const renameParent3 = arrWithDroppable.map((object) => {
      return { ...object, isUpdated: false }
    })
    const renameParent4 = renameParent3.map((object) => {
      return { ...object, isNew: false }
    })
    setArrWith(renameParent4)
  }, [])

  useEffect(() => {
    newArr(nodeData)
  }, [nodeData])

  const handleCallback = (childData) => {
    setNodeDepth(childData)
  }

  const [treeData, setTreeData] = useState([])
  useEffect(() => {
    setTreeData(arrWith)
  }, [arrWith])

  const handleDrop = (newTreeData) => setTreeData(newTreeData)

  const [open, setOpen] = useState(false)
  const [selectedNodes, setSelectedNodes] = useState([])

  const handleSelect = (node) => {
    const item = selectedNodes.find((n) => n.id === node.id)
    if (!item) {
      setSelectedNodes([...selectedNodes, node])
    } else {
      setSelectedNodes(selectedNodes.filter((n) => n.id !== node.id))
    }
  }

  const selectedresult = treeData.concat(selectedNodes)
  const newSelectedResult = [...new Set(selectedresult)]

  console.log(newSelectedResult)

  const handleDelete = (id) => {
    const deleteIds = [id, ...getDescendants(treeData, id).map((node) => node.id)]
    const newTree = treeData.filter((node) => !deleteIds.includes(node.id))
    setTreeData(newTree)
  }

  const handleCopy = (id) => {
    const lastId = getLastId(treeData)
    const targetNode = treeData.find((n) => n.id === id)
    const descendants = getDescendants(treeData, id)
    const partialTree = descendants.map((node) => ({
      ...node,
      id: node.id + lastId,
      parent: node.parent + lastId,
    }))

    setTreeData([
      ...treeData,
      {
        ...targetNode,
        id: targetNode.id + lastId,
      },
      ...partialTree,
    ])
  }

  const handleOpenDialog = () => {
    setOpen(true)
  }

  const handleCloseDialog = () => {
    setOpen(false)
  }
  const [addNewNode, setAddNewNode] = useState(false)
  const [updatedNode, setUpdatedNode] = useState()
  const handleSubmit = (newNode) => {
    setAddNewNode(true)
    setUpdatedNode(false)
    const lastId = getLastId(treeData) + 1

    setTreeData([
      ...treeData,
      {
        ...newNode,
        id: lastId,
      },
    ])

    setOpen(false)
  }

  const handleTextChange = (id, value) => {
    const newTree = treeData.map((node) => {
      if (node.id === id) {
        return {
          ...node,
          text: value,
        }
      }
      return node
    })

    setTreeData(newTree)
  }

  const handleEdit = () => {
    setShowBtn(true)
  }

  const isNewtrue = false
  const ref = useRef(null)
  const handleOpenAll = () => ref.current.openAll()
  const handleCloseAll = () => ref.current.closeAll()

  const variable = location.pathname.substring(location.pathname.lastIndexOf("/") + 1)

  const exportClient = () => {
    URL = Constant.BASE_URL + Constant.PH_CLIENT_EXPORT + "?id=" + variable + "&fileType=" + personName[0] + "&fileName=" + textValue[0]
    window.location.href = URL
    handleClose()
  }

  return (
    <div className="process-hierarchy">
      <Suspense fallback="loading...">
        <ThemeProvider theme={theme}>
          <Navbar />
          <Grid container style={{ marginTop: "10px" }}>
            <Grid item xs={12}>
              <div className="title">
                <h1>
                  {nodeData.description}
                  {showBtn ? <EditOutlinedIcon style={{ width: "15px" }} /> : ""}
                  <Button style={{ float: "right", marginTop: "0" }} variant="contained" onClick={handleClickOpen}>
                    Export File
                  </Button>
                  <Dialog open={openDialog} onClose={handleClose} aria-labelledby="alert-dialog-title" aria-describedby="alert-dialog-description">
                    <DialogTitle id="alert-dialog-title" style={{ fontFamily: "Ubuntu" }}>
                      Export As
                    </DialogTitle>
                    <DialogContent>
                      <DialogContentText id="alert-dialog-description">
                        <Typography style={{ color: "black", fontFamily: "Ubuntu", marginBottom: "10px" }}>Specify a File Name</Typography>
                        <FormControl sx={{ m: 1, width: 451, mt: 3 }}>
                          <OutlinedInput placeholder="Text" onChange={onTextValueChange} />
                        </FormControl>

                        <Typography style={{ color: "black", fontFamily: "Ubuntu" }}>Filetype</Typography>
                        <FormControl sx={{ m: 1, width: 500, mt: 3 }}>
                          <Select
                            displayEmpty
                            value={personName}
                            onChange={handleChange}
                            input={<OutlinedInput />}
                            renderValue={(selected) => {
                              if (selected.length === 0) {
                                return <em>Placeholder</em>
                              }

                              return selected.join(", ")
                            }}
                            MenuProps={MenuProps}
                            inputProps={{ "aria-label": "Without label" }}
                          >
                            <MenuItem disabled value="">
                              <em>Placeholder</em>
                            </MenuItem>
                            {names.map((name) => (
                              <MenuItem key={name} value={name} style={getStyles(name, personName, theme)}>
                                {name}
                              </MenuItem>
                            ))}
                          </Select>
                        </FormControl>
                      </DialogContentText>
                    </DialogContent>
                    <DialogActions>
                      <Button onClick={handleClose}>cancel</Button>
                      <Button onClick={exportClient} autoFocus>
                        Export
                      </Button>
                    </DialogActions>
                  </Dialog>
                </h1>
              </div>
              <div className="rightCol level-container">
                <div style={{ float: "right" }}>
                  <Button variant="outlined" color="primary" onClick={handleOpenAll}>
                    Open All
                  </Button>
                  <Button variant="outlined" color="primary" onClick={handleCloseAll}>
                    Close All
                  </Button>
                  {/* <Button onClick={handleOpen}>Open specific folder(s)</Button>
                <Button onClick={handleClose}>Close specific folder(s)</Button> */}
                </div>
                <div className="legend-blocks">
                  Legend |<span className="level-ones"> Level 1 |</span>
                  <span className="level-two"> Level 2 |</span>
                  <span className="level-three"> Level 3 |</span>
                  <span className="level-four"> Level 4 |</span>
                  <span className="level-five"> Level 5</span>
                </div>

                <DndProvider backend={MultiBackend} options={getBackendOptions()}>
                  <div>
                    {/* <Button onClick={handleOpenDialog} startIcon={<AddIcon />}>
                    Add Node
                  </Button> */}
                    {open && <AddDialog tree={treeData} nodeDepth={nodeDepth} newSelectedResult={newSelectedResult} onClose={handleCloseDialog} onSubmit={handleSubmit} />}
                  </div>

                  <Tree ref={ref} tree={treeData} rootId={"0"} onDrop={handleDrop} initialOpen={"0"} classes={{ container: "wtree", listItem: "parent-listitem" }} sort={false} render={(node, { depth, isOpen, hasChild, onToggle }) => <CustomNode addNewNode={addNewNode} updatedNode={updatedNode} handleEdit={handleEdit} parentCallback={handleCallback} isSelected={!!selectedNodes.find((n) => n.id === node.id)} onSelect={handleSelect} handleCloseDialog={handleCloseDialog} showBtn={showBtn} handleSubmit={handleSubmit} handleOpenDialog={handleOpenDialog} node={node} depth={depth} hasChild={hasChild} isOpen={isOpen} onToggle={onToggle} onTextChange={handleTextChange} onDelete={handleDelete} onCopy={handleCopy} dragPreviewRender={(monitorProps) => <CustomDragPreview monitorProps={monitorProps} />} />} />
                </DndProvider>
              </div>
              <ProcessSave variable={variable} clientValue={props.clientValue} showBtn={showBtn} handleEdit={handleEdit} newSelectedResult={newSelectedResult} />
            </Grid>
          </Grid>
        </ThemeProvider>
      </Suspense>
    </div>
  )
}

export default ClientProcess
`

**Comment by oshi97** [comment]
Also running into a similar issue, where I want to dynamically switch out the tree (NodeModel[]) . My workaround is to manually keep track of open/closed nodes and pass an array into initialOpen instead. The problem with this is that the first render, after switching trees, will always have the nodes closed and then become open. This leads to some UI flashing.

**Comment by oshi97** [comment]
I made a simple example with the workaround and wasn't able to reproduce the flashing so it may be due to application specific details

```ts
import {
  DndProvider,
  getBackendOptions,
  MultiBackend,
  NodeModel,
  RenderParams,
  Tree,
} from "@minoru/react-dnd-treeview";
import { useMemo, useState } from "react";

const ROOT_ID = "root-id";
const trees = [
  [0, 1, 2],
  [10, 11, 12],
];

function App() {
  const [treeIndex, setTreeIndex] = useState(0);
  
  const tree: NodeModel[] = useMemo(() => {
    return trees[treeIndex].map((t) => {
      return {
        id: t,
        parent: t === 0 || t === 10 ? ROOT_ID : t - 1,
        text: t.toString(),
        droppable: true,
      };
    });
  }, [treeIndex]);

  return (
    <div className="App">
      <button onClick={() => setTreeIndex((treeIndex + 1) % 2)}>
        swap trees
      </button>
      <DndProvider backend={MultiBackend} options={getBackendOptions()}>
        <Tree
          tree={tree}
          rootId={ROOT_ID}
          initialOpen={[0,1,2,10,11,12]}
          sort={false}
          insertDroppableFirst={false}
          render={renderNode}
          onDrop={onDrop}
        />
      </DndProvider>
    </div>
  );
}

function onDrop() {}

function renderNode(node: NodeModel, renderParams: RenderParams) {
  console.log(renderParams)
  return <div>{renderParams.isOpen ? 'open' : 'closed'} {node.text}</div>;
}

export default App;
```

**Comment by minop1205** [comment]
The `initialOpen` API is calculated based on the tree data at the time the value changes.
Therefore, if the `initialOpen` value has not changed, or if the tree data does not exist at the time the `initialOpen` value changes, the intended display will not be obtained.

There are three possible ways to set `initialOpen` for lazy-loaded tree data.
Samples of each are provided for your reference.

1. do not render components until tree data is loaded
https://codesandbox.io/s/issue175-sample1-rrl3c8?file=/src/App.jsx

2. open any node using the `open` of `openAll` method after the tree data is loaded
https://codesandbox.io/s/issue175-sample2-0qzr3m?file=/src/App.jsx

3. update the `initialOpen` property when the tree data is loaded
https://codesandbox.io/s/issue175-sample3-v1z1hs?file=/src/App.jsx

**Comment by pivanov** [comment]
Just a hint:
If you use `initialOpen={true}` make sure that your tree items have IDs counting from `0` ... otherwise you should use `initialOpen=[desired-IDs]`. 

Hope this will help! ;)

**Comment by marcvangend** [comment]
Thanks for the awesome library.

In my use case we do not use the opening/closing mechanism at all. Combined with async loading data and dynamically added nodes, I encountered the same problems as described above.

Indeed I was able to keep all nodes at all times using the following method, calling `openAll()` every time the tree data changes (simplified code):

```tsx
const MyTreeComponent = () => {
  const { treeData, setTreeData } = getTreeDataFromApi();
  const treeRef = useRef<TreeMethods>(null);

  useEffect(() => {
    treeRef?.current?.openAll();
  }, [treeData]);

  return (
    <Tree
      ref={treeRef}
      tree={treeData}
      initialOpen={true}
      handleDrop={(newTreeData) => setTreeData(newTreeData)}
      [... other options]
    />
  );
};
```

That said, IMHO it shouldn't be this hard to figure out. It took me a while to even find out why my child nodes weren't rendering as expected. Only then I could start looking for the solution.

The documentation for `initialOpen` says: "If true, all parent nodes will be initialized to the open state." As a developer, I would expect that this means that _all_ droppable nodes are created in the open state, regardless of whether they actually have children, regardless of how and when they are created. At least the documentation could be clearer.
